### ISSUE TITLE
"clusterctl create cluster" command fails to complete

### ISSUE THREAD

**Issue by mshindo** [kubernetes-sigs/cluster-api-provider-vsphere#496]
/kind bug

The problme I'm seeing is that 'clusterctl create cluster' command doesn't finish properly. It waits for Machine management-cluster-controlplane-1 to come up but it never comes up. Finally clusterctl command bails up after 30 mins.

It's similar to the issue #403 but using gcr.io/cnx-cluster-api/vsphere-cluster-api-provider:33dec11d-dirty instead of gcr.io/cluster-api-provider-vsphere/release/manager:v0.4.1 didn't solve my problem.

**What steps did you take and what happened:**

I'm following the instruction described at https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/docs/getting_started.md

**What did you expect to happen:**

clusterctl command completes successfully and the cluster is provisioned on vSphere.

**Anything else you would like to add:**

Here's the output from clusterctl command with -v 5 flag:

    $ clusterctl create cluster   --provider vsphere   --bootstrap-type kind   --cluster ./out/management-cluster/cluster.yaml   --machines ./out/management-cluster/machines.yaml   --provider-components ./out/management-cluster/provider-components.yaml   --addon-components ./out/management-cluster/addons.yaml   --kubeconfig-out ./out/management-cluster/kubeconfig -v 5
    I0811 22:41:10.681847    7076 decoder.go:224] decoding stream as YAML
    I0811 22:41:10.689564    7076 decoder.go:224] decoding stream as YAML
    I0811 22:41:10.691916    7076 createbootstrapcluster.go:27] Preparing bootstrap cluster
    I0811 22:41:10.691936    7076 kind.go:69] Running: kind [create cluster --name=clusterapi]
    I0811 22:42:00.130742    7076 kind.go:72] Ran: kind [create cluster --name=clusterapi] Output: Creating cluster "clusterapi" ...
    • Ensuring node image (kindest/node:v1.15.0) 🖼  ...
    ✓ Ensuring node image (kindest/node:v1.15.0) 🖼
    • Preparing nodes 📦  ...
    ✓ Preparing nodes 📦
    • Creating kubeadm config 📜  ...
    ✓ Creating kubeadm config 📜
    • Starting control-plane 🕹️  ...
    ✓ Starting control-plane 🕹️
    • Installing CNI 🔌  ...
    ✓ Installing CNI 🔌
    • Installing StorageClass 💾  ...
    ✓ Installing StorageClass 💾
    Cluster creation complete. You can now use the cluster with:

    export KUBECONFIG="$(kind get kubeconfig-path --name="clusterapi")"
    kubectl cluster-info
    I0811 22:42:00.130808    7076 kind.go:69] Running: kind [get kubeconfig-path --name=clusterapi]
    I0811 22:42:00.177184    7076 kind.go:72] Ran: kind [get kubeconfig-path --name=clusterapi] Output: /Users/shindom/.kube/kind-config-clusterapi
    I0811 22:42:00.196771    7076 clusterdeployer.go:78] Applying Cluster API stack to bootstrap cluster
    I0811 22:42:00.196795    7076 applyclusterapicomponents.go:26] Applying Cluster API Provider Components
    I0811 22:42:00.197193    7076 clusterclient.go:933] Waiting for kubectl apply...
    I0811 22:42:01.485746    7076 clusterclient.go:962] Waiting for Cluster v1alpha resources to become available...
    I0811 22:42:01.524106    7076 clusterclient.go:975] Waiting for Cluster v1alpha resources to be listable...
    I0811 22:42:01.564910    7076 clusterdeployer.go:83] Provisioning target cluster via bootstrap cluster
    I0811 22:42:01.586463    7076 applycluster.go:36] Creating cluster object management-cluster in namespace "default"
    I0811 22:42:01.603120    7076 clusterdeployer.go:92] Creating control plane machine in namespace "default"
    I0811 22:42:01.614011    7076 applymachines.go:36] Creating machines in namespace "default"
    I0811 22:42:01.633421    7076 clusterclient.go:996] Waiting for Machine management-cluster-controlplane-1 to become ready...
    I0811 22:42:11.642083    7076 clusterclient.go:996] Waiting for Machine management-cluster-controlplane-1 to become ready...
    I0811 22:42:21.641193    7076 clusterclient.go:996] Waiting for Machine management-cluster-controlplane-1 to become ready...
                                                                :
                                                            (snipped)
                                                                :
    I0811 23:11:51.642850    7076 clusterclient.go:996] Waiting for Machine management-cluster-controlplane-1 to become ready...
    I0811 23:12:01.645473    7076 clusterclient.go:996] Waiting for Machine management-cluster-controlplane-1 to become ready...
    I0811 23:12:01.650079    7076 clusterclient.go:996] Waiting for Machine management-cluster-controlplane-1 to become ready...
    I0811 23:12:01.661690    7076 createbootstrapcluster.go:36] Cleaning up bootstrap cluster.
    I0811 23:12:01.661713    7076 kind.go:69] Running: kind [delete cluster --name=clusterapi]
    I0811 23:12:03.118023    7076 kind.go:72] Ran: kind [delete cluster --name=clusterapi] Output: Deleting cluster "clusterapi" ...
    F0811 23:12:03.118065    7076 create_cluster.go:61] unable to create control plane machine: timed out waiting for the condition

**Environment:**

- Cluster-api-provider-vsphere version: v0.4.1
- Kubernetes version: (use `kubectl version`):  client: v1.15.2, server: v1.15.0
- OS (e.g. from `/etc/os-release`): MacOS Mojave (10.14.6)

**Comment by mshindo** [comment]
Sorry, it was not a bug. I noticed that there was a warning in the output of `kubectl describe machine` command as follows:

    Warning  CreateFailure  34s                vsphere-controller  error generating user data for "default/management-cluster/management-cluster-controlplane-1": unable to get folder for "default/management-cluster/management-cluster-controlplane-1": path 'vm' resolves to multiple folders

I was using `export VSPHERE_FOLDER='vm'` in the envvar.txt as shown in https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/docs/getting_started.md and provisioned the corresponding 'vm' folder on the target vSphere. As the warning indicates, this was the root of the problem. I changed the folder name to something else and it started to work properly. It was my trivial mistake but it may not be a good idea to use a folder named 'vm' as an example in the Getting Started document.

Please close this Issue.

**Comment by figo** [comment]
@mshindo  thank you for reporting the issue, i will looking into why folder `vm` does not work.

**Comment by tkrausjr** [comment]
@figo  I believe this is expected behavior of govmomi and is a result of the govmomi finder method being used which will search the inventory tree for all occurrences of the object(folder) and if its found in multiple locations the function will return an error. Code is here https://github.com/vmware/govmomi/blob/dd082db295a7575ad1a8c00690bf66a508b9bb3f/find/finder.go#L1047-L1058

Explanation of default behavior when full path with slashes is provided vs just name.
https://godoc.org/github.com/vmware/govmomi/find

To avoid it you can use the full path to the folder ie. /Datacenter/vm

**Comment by akutz** [comment]
Hi @figo / @mshindo / @tkrausjr,

Thomas is correct. The reason the docs use `vm` is because it's the name of the root folder. There's nothing wrong with that, but if there's any ambiguity, it will fail. There's nothing we can do to prevent that -- it's on the operator to provide additional specificity. 

I literally just answered https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/481 by referencing this comment https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/403#issuecomment-510276822.

Basically it's on the user to ensure that the resource paths are qualified whenever required. @tkrausjr is actually working on #429 to validate the cluster and machine spec information. Thomas, is there anything you'd like to add with regards to validation?

**Comment by tkrausjr** [comment]
Nothing to add, waiting on some feedback on #429 but this work will allow the checks to be done before the actual actuators kick in so that the user can fail faster and have a clearer understanding of what the issue was.

**Comment by akutz** [comment]
Hi @figo,

Do you need help looking into this? What's the status? Thanks!

**Comment by akutz** [comment]
Hi @mshindo,

Can you possible confirm whether or not this is still an issue and follow up with @figo? Thank you!
